proposals/20230316-oci-content-source.md

```diff
@@ -0,0 +1,284 @@
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.wasm.content.layer.v1+wasm",
+      "digest": "sha256:b36aa5d0111a488937361fdb35432510d50675a11d566c5d9e82a147fb9ff552",
+      "size": 2087464,
+      "annotations": {
+        "org.opencontainers.image.title": "my-component"
+      }
+    }
+  ]
+}
+```
+
+The following is an example of the configuration structure referenced in the preceding image manifest.
+```json
+{
+  "mediaType": "application/vnd.wasm.component.config.v1+json",
+  "architecture": "wasm32",
+  "os": "wasi"
+}
+```
+
+#### Image Manifest for Interface Components
+The following is an example image manifest for a component containing a configuration structure and a layer containing the `my-component-interface.wasm` binary.
+```json
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "artifactType": "application/vnd.wasm.component.interface.v1",
+  "config": {
+    "mediaType": "application/vnd.wasm.component.config.v1+json",
+    "digest": "sha256:c71d239df91726fc519c6eb72d318ec65820627232b2f796219e87dcf35d0ab4",
+    "size": 331
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.wasm.content.layer.v1+wasm",
+      "digest": "sha256:dcf07c6bb395d6e1d40505b77e70af04f5fae0d54c9573fd379c1e7355a18cf3",
+      "size": 2087464,
+      "annotations": {
+        "org.opencontainers.image.title": "my-component-interface"
+      }
+    }
+  ]
+}
+```
+
+The following is an example of the configuration structure referenced in the preceding image manifest.
+```json
+{
+  "mediaType": "application/vnd.wasm.component.config.v1+json",
+  "architecture": "wasm32",
+  "os": "wasi"
+}
+```
+
+#### Image Manifest for Bundled Components
+The following is an example image manifest for a component containing a configuration structure, a layer containing the `my-bundled-component.wasm` binary, and a data layer containing some a static asset.
+```json
+{
+  "schemaVersion": 2,
+  "mediaType": "application/vnd.oci.image.manifest.v1+json",
+  "artifactType": "application/vnd.wasm.component.bundled.v1",
+  "config": {
+    "mediaType": "application/vnd.wasm.component.config.v1+json",
+    "digest": "sha256:105ab3237b4f0d885700892a0f4b3482d1146dff27c88d46f02b8bd7ef67c3de",
+    "size": 331
+  },
+  "layers": [
+    {
+      "mediaType": "application/vnd.wasm.content.layer.v1+wasm",
+      "digest": "sha256:2e94e0582fb925e89515435513496819dc8f364f2da400059a64d6d1412ca2ad",
+      "size": 2087464,
+      "annotations": {
+        "org.opencontainers.image.title": "my-bundled-component"
+      }
+    },
+    {
+      "mediaType": "application/vnd.wasm.content.layer.v1+data",
+      "digest": "sha256:8c69a84ec5adec97e47d4250410a7689046762aaa8e89f82ddbb4a89acb7388e",
+      "size": 96
+    }
+  ]
+}
+```
+
+The following is an example of the configuration structure referenced in the preceding image manifest.
+```json
+{
+  "mediaType": "application/vnd.wasm.component.config.v1+json",
```

Comment:
Expanding on what I was saying above, I don't think we should consider this file "component configuration", but rather some sort of "wasm app" configuration.  Once `wasi-virt` is built and working, I think there shouldn't even need to *be* a concept of "component configuration": the contents of this configuration file should go *into the component*: files go into data segments in a component generated by `wasi-virt` that *bundles* the original component (so that the new outer component does not import `wasi:filesystem/types` at all).  Similarly, environment variables can be set by having `wasi-virt` virtualize `wasi:cli/environment`.  The large win from doing it this way is that the output of `wasi-virt` is *still a component* that can be further composed and manipulated by downstream component tooling (enabling virtual platform layering).  In the meantime, it makes total sense to use this config file as a stopgap (or maybe in perpetuity if folks are wanting to do other app-level configuration stuff, which I've heard); I just want to name it appropriately so that "component" really does mean "component" (and nothing else).